### PR TITLE
Now, `dpo.py` matches `dpo_tune_cache.py` almost perfectly on the single GPU experiments

### PR DESCRIPTION
**Pull request by finbarrtimbers**
[Colab](https://colab.research.google.com/drive/19olnAYsNSfp_BkmdkrGqgmNhdEY8Uca-#scrollTo=-DucPa2CVnlc) with plots, particularly loss curve: 

Screenshot 2026-02-01 at 9 06 23 AM

Changes: 

1. In `HFDataLoader`, switched to match the `dpo_tune_cache.py` iteration order by using `torch.randperm` with seeded `torch.Generator`
2. Fix attention masking in the `dpo.py` forwards pass by converting padded batches to the packed format. We now pass `doc_lens` and `max_doc_lens`, enabling proper intra-document attention masking. 
3. Fix weight loading order by loading after paralellization (this was previously re-initializing the weights, whoops). 
4. Use 0-based epoch numbers to match `dpo_tune_cache.py`. 
5. Use `Duration.steps` instead of `Duration.epochs` in `dpo.py` so we do `max_train_steps` training steps. 

Changes not related to making DPO match: 

1. Fixed the `token_count` metric in `dpo_tune_cache.py` to undo the mean reduction. 
2. Make data ordering reproducible by using a seeded `RandomSampler` with `torch.Generator().manual_seed(seed)`. 

# Known areas of deviation between the two: 

1. `dpo.py` sets `drop_last=True` while `dpo_tune_cache.py` doesn't. 
2. `dpo.py` accumulates gradients using a token-weighted mean, while `dpo_tune_cache.py` uses uniform weighting. 

We think that `dpo.py` does things correctly. 

---

> [!NOTE]
> **Medium Risk**
> Changes core DPO training semantics (data order, masking, checkpoint weight loading, and step scheduling), which can materially affect convergence and reproducibility; scope is contained to DPO paths and covered by updated GPU tests.
> 
> **Overview**
> Makes `dpo.py`’s single-GPU behavior match `dpo_tune_cache.py` more closely by (1) switching `HFDataLoader` shuffling to seeded `torch.randperm` for reproducible ordering, (2) fixing OLMo-core attention masking by repacking padded batches into `doc_lens`/`max_doc_lens` format in `dpo_utils` forwards, and (3) reloading HF checkpoint weights *after* `TransformerTrainModule` parallelization to avoid accidental re-init.
> 
> Also updates training control to be step-based (`max_train_steps` via `Duration.steps` + scheduler step count), forces `drop_last=True` for DPO data/cache loaders, makes `dpo_tune_cache.py` sampling reproducible with a seeded `RandomSampler`, and corrects its `token_count` aggregation after mean reduction; tests and debug script are adjusted accordingly.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c5cb1bbea8382db848a3077463c8c62cc89ca366. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: open_instruct/test_dpo_utils_gpu.py [hunk 1]
```diff
@@ -80,7 +80,10 @@ def __init__(self, vocab_size: int = 1000):
         self.embed = torch.nn.Embedding(vocab_size, 64)
         self.linear = torch.nn.Linear(64, vocab_size)
 
-    def forward(self, input_ids: torch.Tensor) -> torch.Tensor:
+    def forward(
+        self, input_ids: torch.Tensor, doc_lens: torch.Tensor | None = None, max_doc_lens: list[int] | None = None
+    ) -> torch.Tensor:
+        del doc_lens, max_doc_lens
```
Comment (hamishivi) [inline open_instruct/test_dpo_utils_gpu.py]: why del? Weird side effect.

Comment (finbarrtimbers) [reply open_instruct/test_dpo_utils_gpu.py]: Removed!

**Review by hamishivi — APPROVED** [review]
